modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDmlTest.java

```diff
@@ -672,6 +675,32 @@ public void testDeleteUsingCompositePk() {
         assertQuery("SELECT d FROM test").returnNothing().check();
     }
 
+    @Test
+    public void testTransactionFailsToCommitWhenDetectsSchemaChange() {
+        sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        CompletableFuture<Void> upsertFut = CompletableFuture.runAsync(() -> {
+            sql("INSERT INTO TEST (id, val0) VALUES (?, ?)", 1, 1);
```

Comment:
Do you mean this statement should be run before "ALTER table?
There is no such guarantee, as async task may starts after "ALTER TABLE"

Maybe an explicit transaction, which spans over both INSERTS should be started?

Reply:
Fixed.